core/base/workspace_aliases.hpp

```diff
@@ -0,0 +1,279 @@
+#if defined(__CUDACC__)
+#define GKO_DEVICE_ERROR_TYPE cudaError_t
+#define GKO_DEVICE_ERROR_INVALID cudaErrorInvalidValue
+#define GKO_DEVICE_NO_ERROR cudaSuccess
+#elif defined(__HIPCC__)
+#define GKO_DEVICE_ERROR_TYPE hipError_t
+#define GKO_DEVICE_ERROR_INVALID hipErrorInvalidValue
+#define GKO_DEVICE_NO_ERROR hipSuccess
+#else
+#define GKO_DEVICE_ERROR_TYPE int
+#define GKO_DEVICE_ERROR_INVALID 1
+#define GKO_DEVICE_NO_ERROR 0
+#endif
+
+
+// This code is a modified version of the code from CCCL
+// (https://github.com/NVIDIA/cccl) (cub/detail/temporary_storage.cuh and
+// cub/temporary_storage.cuh), made available through the Apache-2.0 and BSD-3
+// licenses. See ABOUT-LICENSING.md for more details.
+// The main changes: the namespaces, the object/class
+// names and some simplifications that adhere to Ginkgo's code conventions
+
+
+namespace gko {
+namespace detail {
+
+
+template <int num_allocs, int align_bytes = 8>
+GKO_ATTRIBUTES GKO_INLINE GKO_DEVICE_ERROR_TYPE create_workspace_aliases(
+    void* workspace_ptr, size_t& num_bytes, void* (&allocations)[num_allocs],
+    const size_t (&allocation_sizes)[num_allocs])
+{
+    constexpr int align_mask = ~(align_bytes - 1);
+
+    // Compute exclusive prefix sum over allocation requests
+    size_t allocation_offsets[num_allocs];
+    size_t bytes_needed = 0;
+    for (int i = 0; i < num_allocs; ++i) {
+        size_t allocation_bytes =
+            (allocation_sizes[i] + align_bytes - 1) & align_mask;
+        allocation_offsets[i] = bytes_needed;
+        bytes_needed += allocation_bytes;
+    }
+    bytes_needed += align_bytes - 1;
+
+    // Check if the caller is simply requesting the size of the storage
+    // allocation
+    if (!workspace_ptr) {
+        num_bytes = bytes_needed;
+        return GKO_DEVICE_NO_ERROR;
+    }
+
+    // Check if enough storage provided
+    if (num_bytes < bytes_needed) {
+        return GKO_DEVICE_ERROR_INVALID;
+    }
+
+    // Alias
+    workspace_ptr =
+        (void*)((size_t(workspace_ptr) + align_bytes - 1) & align_mask);
+    for (int i = 0; i < num_allocs; ++i) {
+        allocations[i] =
+            static_cast<char*>(workspace_ptr) + allocation_offsets[i];
+    }
+
+    return GKO_DEVICE_NO_ERROR;
+}
+
+
+class slot;
+
+template <typename T>
+class alias;
+
+template <int num_slots, int align_bytes>
+class layout;
+
+/**
+ * Represents a slot of memory which can be aliased. See @alias and @layout on
+ * how this can be used.
+ */
+class slot {
+    template <typename T>
+    friend class alias;
+
+    template <int, int>
+    friend class layout;
+
+public:
+    slot() = default;
+
+    /**
+     * @brief Returns an array of type @p T and length @p num_elems
+     */
+    template <typename T>
+    GKO_ATTRIBUTES alias<T> create_alias(std::size_t num_elems = 0);
+
+private:
+    GKO_ATTRIBUTES void set_bytes_required(std::size_t new_size)
+    {
+        size_ = max(size_, new_size);
```

Comment:
this causes an error in our no-circular-deps test, because nothing provides the `max` function. I think it should be fine to use `std::max`. It should even be available in device code, since it's `constexpr` since c++14. If that is somehow not the case, then removing the `GKO_ATTRIBUTES` should be fine, since this functionality is not used in device codes anyway.